nav2_costmap_2d/include/nav2_costmap_2d/costmap_filters/keepout_filter.hpp

```diff
@@ -0,0 +1,83 @@
+/*********************************************************************
+ *
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020 Samsung Research Russia
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the <ORGANIZATION> nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Alexey Merzlyakov
+ *********************************************************************/
+
+#ifndef NAV2_COSTMAP_2D__KEEPOUT_FILTER_HPP_
+#define NAV2_COSTMAP_2D__KEEPOUT_FILTER_HPP_
+
+#include "nav2_costmap_2d/costmap_filters/costmap_filter.hpp"
+
+#include <string>
+#include <memory>
+#include <mutex>
+
+#include "rclcpp/rclcpp.hpp"
+#include "nav_msgs/msg/occupancy_grid.hpp"
+#include "nav2_msgs/msg/costmap_filter_info.hpp"
+
+namespace nav2_costmap_2d
+{
+
+class KeepoutFilter : public CostmapFilter
+{
+public:
+  KeepoutFilter();
+
+  void initializeFilter(
+    const std::string & filter_info_topic);
+
+  void process(
+    nav2_costmap_2d::Costmap2D & master_grid,
+    int min_i, int min_j, int max_i, int max_j,
+    const geometry_msgs::msg::Pose2D & pose);
+
+  void resetFilter();
+
+private:
+  void filterInfoCallback(const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg);
+  void maskCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg);
+
+  std::string filter_info_topic_, mask_topic_;
```

Comment:
For `filter_info_topic_` and `mask_topic_` I would like agree that this is better to keep in basic parent class CostmapFilter. For the last one `mask_costmap_` - this is the object with Costmap2D type, which is used only by KeepoutFilter. I do not think, we need to bind CostmapFilter with specific filter realization, rather to keep it more or less abstract. For SpeedFilter, for example, I plan to directly working with OccupancyGrid maps instead of copying whole info into Costmap2D (this is not a fixed yet, but I think we can see the principal from here). So, I would like to remain  `mask_costmap_` to belonging to KeepoutFilter class.

Reply:
OK. I think that will be used by most every one but I can understand that.